paf-mvp-core-js/src/express/node.ts

```diff
@@ -85,27 +87,76 @@ export class Node implements INode {
   /**
    * Returns a handler that handles errors
    * @param endPointName
-   * @protected
    */
-  protected handleErrors(endPointName: string): ErrorRequestHandler {
-    return (err: ClientNodeError, req: Request, res: Response, next: NextFunction) => {
+  handleErrors(endPointName: string): ErrorRequestHandler {
+    return (err: NodeError, req: Request, res: Response, next: NextFunction) => {
       // TODO next step: define a common logging format for errors (on 1 line), usable for monitoring
       this.logger.Error(endPointName, err);
+      // For redirect endpoints, if the provided redirect url is invalid, try to redirect back to referer
+      if (err.type === NodeErrorType.INVALID_RETURN_URL) {
+        this.redirectToReferer(req, res, endPointName);
+      }
     };
   }
 
+  private redirectToReferer(req: Request, res: Response, endPointName: string) {
+    const referer = req.header('referer');
+    this.logger.Warn(endPointName, `Redirecting to referer: ${referer} ...`);
+    try {
+      const redirectUrl = new URL(referer);
+      httpRedirect(res, redirectUrl.toString());
+    } catch (e) {
+      this.logger.Error(endPointName, e);
+    }
+  }
+
   /**
-   * Build an handler that validate the body of the Request
+   * Build a handler that validates the body of the Request
    * against a JSON schema.
-   * @returns the built hander
+   * @returns the built handler
    */
-  protected buildJsonBodyValidatorHandler(jsonSchema: string): RequestHandler {
+  buildJsonBodyValidatorHandler(jsonSchema: string): RequestHandler {
     return (req: Request, res: Response, next: NextFunction) => {
       const validation = this.jsonValidator.validate(jsonSchema, req.body as string);
       if (!validation.isValid) {
         const details = validation.errors.map((e) => e.message).join(' - ');
-        const error: ClientNodeError = {
-          type: ClientNodeErrorType.INVALID_JSON_BODY,
+        const error: NodeError = {
+          type: NodeErrorType.INVALID_JSON_BODY,
+          details,
+        };
+        res.status(400);
+        res.json(error);
+        next(error);
+      } else {
+        next();
+      }
+    };
+  }
+
+  /**
+   * Builds a handler that validates the query string
+   * against a JSON schema.
+   * @returns the built handler
+   */
+  buildQueryStringValidatorHandler(jsonSchema: string): RequestHandler {
+    return (req: Request, res: Response, next: NextFunction) => {
+      const data = req.query[QSParam.paf] as string | undefined;
+      const decodedData = data ? decodeBase64(data) : undefined;
+      if (!decodedData) {
+        const error: NodeError = {
+          type: NodeErrorType.INVALID_QUERY_STRING,
+          details: `Received Query string: ${data} is not a valid Base64 string`,
```

Comment:
The error message should be a bit awkward if there is no query string parameter.
You can add an error case above or make the error message more generic.

Reply:
done